styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -354,31 +376,85 @@ public void updateNextNaturalTrigger(WorkflowId workflowId, Instant instant) thr
     final QueryResults<Entity> results = datastore.run(activeStatesQuery);
     while (results.hasNext()) {
       final Entity entity = results.next();
       final long counter = entity.getLong(PROPERTY_COUNTER);
       final WorkflowInstance instance = parseWorkflowInstance(entity);
+      final PersistentWorkflowInstanceStateBuilder persistentState =
+          PersistentWorkflowInstanceState.builder()
+              .counter(counter);
+
+      // TODO: always read these state fields when all active state entities have been migrated
+      if (entity.contains(PROPERTY_STATE)) {
+        final State state = State.valueOf(entity.getString(PROPERTY_STATE));
+        final long timestamp = entity.getLong(PROPERTY_STATE_TIMESTAMP);
+
+        final StateData data = StateData.newBuilder()
+            .tries((int) entity.getLong(PROPERTY_STATE_TRIES))
+            .consecutiveFailures((int) entity.getLong(PROPERTY_STATE_CONSECUTIVE_FAILURES))
+            .retryCost(entity.getDouble(PROPERTY_STATE_RETRY_COST))
+            .trigger(this.<String>readOpt(entity, PROPERTY_STATE_TRIGGER_TYPE).map(type ->
+                TriggerUtil.trigger(type, entity.getString(PROPERTY_STATE_TRIGGER_ID))))
+            .messages(OBJECT_MAPPER.<List<Message>>readValue(entity.getString(PROPERTY_STATE_MESSAGES),
+                new TypeReference<List<Message>>() { }))
+            .retryDelayMillis(readOpt(entity, PROPERTY_STATE_RETRY_DELAY_MILLIS))
+            .lastExit(readOpt(entity, PROPERTY_STATE_LAST_EXIT))
+            .executionId(readOpt(entity, PROPERTY_STATE_EXECUTION_ID))
+            .executionDescription(readOptJson(entity, PROPERTY_STATE_EXECUTION_DESCRIPTION,
+                ExecutionDescription.class))
+            .build();
+
+        persistentState
+            .state(state)
+            .data(data)
+            .timestamp(Instant.ofEpochMilli(timestamp));
+      }
 
-      mapBuilder.put(instance, counter);
+      mapBuilder.put(instance, persistentState.build());
     }
 
     return mapBuilder.build();
   }
 
-  void writeActiveState(WorkflowInstance workflowInstance, long counter) throws IOException {
+  void writeActiveState(WorkflowInstance workflowInstance, PersistentWorkflowInstanceState state)
+      throws IOException {
     storeWithRetries(() -> {
       final Key key = activeWorkflowInstanceKey(workflowInstance);
-      final Entity entity = Entity.newBuilder(key)
+      final Entity.Builder entity = Entity.newBuilder(key)
           .set(PROPERTY_COMPONENT, workflowInstance.workflowId().componentId())
           .set(PROPERTY_WORKFLOW, workflowInstance.workflowId().id())
           .set(PROPERTY_PARAMETER, workflowInstance.parameter())
-          .set(PROPERTY_COUNTER, counter)
-          .build();
+          .set(PROPERTY_COUNTER, state.counter());
+
+      // TODO: always write these fields when event-only replay tests have been removed
+      if (state.state() != null) {
```

Comment:
Where in WFI lifecycle does this become nonnull? I kind of expected the state to be unconditionally _written_ to Datastore here - then it would be clear to me that in 24 hours all the active states are dark-initialized in Datastore alright, but now I'm a bit confused.

Reply:
During normal operations the state should always be non-null. 

Like the TODO mentions, some tests write `PersistentWorkflowInstanceState` instances with null `state` though in order to test scenarios where we fall back to event replay from bigtable due to state not being in datastore.

Reply:
Oh so it's only null in tests...